Week04

safiweb opened the pull request:
# Adding Artist, Band, Venue struct types, interfaces implementation, and methods

This PR introduces _Artist_, _Band_, and _Venue_ struct.  _Artist_ and _Band_ struct implement the _Entertainer_ interface and one of either _Setuper_ or _Teardowner_ interface. The _Venue_ struct implements _Entertain_ method that takes the number of audience members and a list of _Entertainer_.

The PR also adds unit test functions for all methods implemented by the various struct types and interfaces.

## Changes Summary
- Added new file artist.go - adds Artist struct type and its interface implementation
- Added new file artist_test.go - adds the unit test functions for the Artist struct type and its interface implementation
- Added new file band.go - adds Band struct type and its interface implementation
- Added new file band_test.go - adds the unit test functions for the Band struct type and its interface implementation
- Added new file venue.go - adds Venue struct type and its interface implementation
- Added new file venue_test.go - adds the unit test functions for the Venue struct type and its interface implementation
- Added new file interfaces.go - add the Entertainer, Setuper, and Teardowner interfaces.
- Miscellaneous - formatting code

## Difficulties:
- Totally misread the assignment requirements(watched the discussion video twice :)) and implemented something differently. I had thought the assignment was constrained to only the provided Venue struct type. I have redone my work to reflect expected requirements.

- In testing, all methods that call fmt.Fprintf(v.Log,......) on success scenarios, I was getting panic error. I had to catch the fmt.Fprintf output via the &bytes.Buffer{} String method.

```
buff := bytes.Buffer{}
tt.v.Log = &buff
..........
..........
got := buff.String()
```

## Surprises

- The &bytes.Buffer{} implements io.Writer and can use it String method to catch the log output.

safiweb committed:
Week 4 assignment set up

safiweb committed:
Setting up Venue type + various interfaces and methods

safiweb committed:
Adding unit test functions

markbates commented on week04/assignment04/assignment04.go:
```diff
@@ -0,0 +1,85 @@
+		return fmt.Errorf("%v cannot perform for 0 audience", p.Name())
+	}
+	return nil
+}
```
The `Venue` should not implement the `Entertainer` interface. You are supposed to create two new types that implement the different interfaces.

safiweb replied:
ok, totally missed that. Redone my work :)

markbates commented on week04/assignment04/assignment04.go:
```diff
@@ -0,0 +1,85 @@
+	}
+
+	return nil
+}
```
Again, the `Venue` should not implement `Setuper`. One of your `Entertainer` implementations should implement the `Setuper` interface, the other should implement `Teardowner`.

markbates commented on week04/assignment04/assignment04_test.go:
```diff
@@ -0,0 +1,142 @@
+	{"success", 20, []Entertainer{
+		Venue{Audience: 5, Log: os.Stdout},
+	}, nil},
+}
```
Using table driven tests is excellent! However, creating global variables like this is generally considered bad practice in Go in regular code and tests. Test cases belong inside the test and are **never** globally.

markbates reviewed (commented):
You are currently not implementing the interfaces as described in the assignment. Please re-read the assignment and make the necessary changes and re-submit when done.

I did not give you feedback on your tests, other than to say never use global variables. Because your implementation is incorrect the tests will need to be updated. I will provide detailed feedback on them when you re-submit.

markbates commented on week04/assignment04/assignment04_test.go:
```diff
@@ -0,0 +1,142 @@
+		v    Venue
+		act  Entertainer
+	}{
+		{"success", Venue{Audience: 10, Log: os.Stdout}, Venue{}},
```
You don't want to use `os.Stdout` for this because you can't make assertions easily against. Use a type such as `&bytes.Buffer{}` which implements `io.Writer` but can easily be asserted against. For example, it has a `String() string` method that can be used to get everything written to the log.

safiweb committed:
Refactoring code

- Added Artist + Band struct type
- Reorganized code files
- Added new unit test functions

markbates commented on week04/assignment04/artist.go:
```diff
@@ -0,0 +1,32 @@
+import "fmt"
+
+type Artist struct {
+	Artistname string
```
In Go, we use PascalCase for naming, `ArtistName`. However, we also try not to be redundant. This field does not need the `Artist` prefix.

```suggestion
	Name string
```

safiweb replied:
I was getting a "Field and method with the same name" error when I use Name as the field name.

safiweb committed:
Changes to variable naming

safiweb committed:
Github Actions updated
